### PR TITLE
[IMP] Modifier, Link : Add a preserve property , better behaviors of Link format

### PR DESCRIPTION
**Pull request by sebgeelen**
This property is used to determine if a modifier should be kept after a break line or not.

see issue #187

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,13 +1,13 @@
+# dotfiles
+.*
```
Comment (dmo-odoo) [inline .gitignore:2]: Smart ! Then I guess you could have removed `.cache` from the list... ;-) That being said, it would be better if this was in a separate commit as it is not related to the `Modifier` improvement here. Same PR is ok but not same commit. I suggest `[IMP] Core: improve gitignore` or something along these lines.

File: packages/core/src/Modifier.ts [hunk 2]
```diff
@@ -8,6 +8,7 @@ export interface Modifier {
     constructor: ModifierConstructor & this;
 }
 export class Modifier {
+    public preserve: boolean = true;
```
Comment (dmo-odoo) [inline packages/core/src/Modifier.ts:11]: `public` is the default, we have chosen to omit it when defining class properties. The only place where we use it is in constructor parameters.

File: packages/plugin-char/src/Char.ts [hunk 3]
```diff
@@ -47,7 +48,12 @@ export class Char<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
         const range = params.context.range;
         const text = params.text;
         const inline = this.editor.plugins.get(Inline);
-        const modifiers = inline.getCurrentModifiers(range);
+        let modifiers = inline.getCurrentModifiers(range);
+
+        const cleanModifier = range.start.previousSibling() === undefined || range.start.previousSibling().is(LineBreakNode);
```
Comment (dmo-odoo) [inline packages/plugin-char/src/Char.ts:53]: I generally avoid to write `undefined` unless in rare cases where it can be a somewhat valid value.

In this particular case I would have done:
```typescript
const previousSibling = range.start.previouspreviousSibling();
if (!previousSibling || previousSibling.is(LineBreakNode)) {
```

In the version with `undefined`, you have to "translate" in your brain the fact that `previousSibling` being `undefined` means that there is no previous sibling. The suggested code is more idiomatic IMO because it reads exactly what it means: "if there is no previous sibling or it is a LineBreakNode". You don't need to wrap your brain about anything in particular.

Comment (sebgeelen) [reply packages/plugin-char/src/Char.ts:53]: Agree.

File: packages/core/src/Modifier.ts (hunk 2 above)
Comment (sebgeelen) [reply packages/core/src/Modifier.ts:11]: I'll rewrite the commit ;)

File: packages/plugin-char/src/Char.ts [hunk 4]
```diff
@@ -47,7 +48,12 @@ export class Char<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
         const range = params.context.range;
         const text = params.text;
         const inline = this.editor.plugins.get(Inline);
-        const modifiers = inline.getCurrentModifiers(range);
+        let modifiers = inline.getCurrentModifiers(range);
+
+        const previousSibling = range.start.previousSibling();
+        if(!previousSibling || previousSibling.is(LineBreakNode)) {
```
Comment (Zinston) [inline packages/plugin-char/src/Char.ts:54]: Here's the unlinted culprit :-) `if (`

File: packages/core/src/Modifier.ts [hunk 5]
```diff
@@ -8,6 +8,7 @@ export interface Modifier {
     constructor: ModifierConstructor & this;
 }
 export class Modifier {
+    preserve: boolean = true;
```
Comment (Zinston) [inline packages/core/src/Modifier.ts:11]: Is the typing necessary? Doesn't typescript recognize that setting the property to `true` means its type is boolean? Or does it think the type is `true` then?

Comment (dmo-odoo) [reply packages/core/src/Modifier.ts:11]: No it's not necessary. I think our prettier config fixes that kind of things automatically.

File: packages/plugin-char/src/Char.ts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import { JWPlugin, JWPluginConfig } from '../../core/src/JWPlugin';
 import { CharNode } from './CharNode';
+import { LineBreakNode } from '../../plugin-linebreak/src/LineBreakNode';
```
Comment (Zinston) [inline packages/plugin-char/src/Char.ts:3]: Be careful when referencing plugins in other plugins. This one should not be mandatory for `Char` to work.

File: packages/plugin-link/src/Link.ts [hunk 7]
```diff
@@ -30,7 +30,7 @@ export class Link<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
             node = link;
         }
         const format = node.is(InlineNode) && node.modifiers.find(LinkFormat);
-        return link instanceof AbstractNode ? !!format : format === link;
+        return link instanceof AbstractNode ? !!format : link.isSameAs(format);
```
Comment (Zinston) [inline packages/plugin-link/src/Link.ts:33]: :+1:

File: packages/plugin-dom-layout/src/ui/DomLayoutEngine.ts [hunk 8]
```diff
@@ -120,8 +120,12 @@ export class DomLayoutEngine extends LayoutEngine {
...
      * @override
      */
-    async add(componentId: ComponentId, zoneId: ZoneIdentifier): Promise<VNode[]> {
-        const nodes = await super.add(componentId, zoneId);
+    async add(
+        componentId: ComponentId,
+        zoneId: ZoneIdentifier,
+        params?: Record<string, any>,
+    ): Promise<VNode[]> {
+        const nodes = await super.add(componentId, zoneId, params);
```
Comment (Zinston) [inline packages/plugin-dom-layout/src/ui/DomLayoutEngine.ts:128]: This whole thing is cool but should be split into its own commit.

File: packages/plugin-link/src/Link.ts [hunk 9]
```diff
@@ -73,32 +73,54 @@ export class Link<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
...
+    }
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
 
-        const sameLink: Typeguard<InlineNode> = Link.isLink.bind(Link, link);
-        for (const inline of [node, ...node.adjacents(sameLink)]) {
-            inline.modifiers.remove(link);
+    _removeLinkOnNodes(nodes: VNode[]): void {
```
Comment (Zinston) [inline packages/plugin-link/src/Link.ts:120]: doc please :-)

File: packages/plugin-link/src/Link.ts [hunk 10]
```diff
@@ -73,32 +73,54 @@ export class Link<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
...
+        for (const node of nodes) {
+            const link = node.modifiers.find(LinkFormat);
+            node.modifiers.remove(link);
```
Comment (Zinston) [inline packages/plugin-link/src/Link.ts:123]: `node.modifiers.remove(LinkFormat)` should suffice :-)

File: packages/plugin-xml/src/Attributes.ts [hunk 11]
```diff
@@ -5,6 +5,7 @@ import { ClassList } from './ClassList';
 export class Attributes extends Modifier {
     private _record: Record<string, string> = {};
     style = new CssStyle();
+    preserve = false;
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply packages/plugin-xml/src/Attributes.ts:8]: You tell me. 

@dmo-odoo or you ( I don't remember) told me to put the preserve false on attributes. 
I can't say I understand why.

File: packages/plugin-layout/src/LayoutEngine.ts [hunk 12]
```diff
@@ -8,7 +8,7 @@ export type DomZonePosition = 'after' | 'prepend' | 'append' | 'replace';
 
 export interface ComponentDefinition {
     id: ComponentId;
-    render: (editor: JWEditor) => Promise<VNode[]>;
+    render: (editor: JWEditor, params?: Record<string, any>) => Promise<VNode[]>;
```
Comment (dmo-odoo) [inline packages/plugin-layout/src/LayoutEngine.ts:11]: Your linter let you do that ? That's weird because we're supposed to have a shared configuration and ours does not let us write `any`. We usually write `{}` instead. Like `any` it matches everything, but it will generate an error if you try to use it in any meaningful way.

With any, you can do that:
```ts
var a: Any = { a: 'a' };
console.log(a.b); // no warning
```

With `{}`, you can't do that:
```ts
var a: {} = { a: 'a' };
console.log(a.b); // warning: `{}` does not have a `b` property
```

File: packages/plugin-link/src/Link.ts [hunk 13]
```diff
@@ -65,8 +65,8 @@ export class Link<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
         components: [
             {
                 id: 'link',
-                async render(): Promise<OwlNode[]> {
-                    return [new OwlNode({ Component: LinkComponent, props: {} })];
+                async render(editor: JWEditor, params?: Record<string, any>): Promise<OwlNode[]> {
```
Comment (dmo-odoo) [inline packages/plugin-link/src/Link.ts:68]: @Zinston @Goaman Shall we name the component params as `props` or do you think it might get confusing if people use them as something else than props in their component's `render` method ?

File: packages/plugin-link/src/Link.ts [hunk 14]
```diff
@@ -132,23 +138,41 @@ export class Link<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
...
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
 
-        const sameLink: Typeguard<InlineNode> = Link.isLink.bind(Link, link);
-        for (const inline of [node, ...node.adjacents(sameLink)]) {
-            inline.modifiers.remove(link);
+    _removeLinkOnNodes(nodes: VNode[]): void {
+        // remove all link modifiers on the provided nodes
```
Comment (dmo-odoo) [inline packages/plugin-link/src/Link.ts:172]: Can I ask you to capitalize the first letters of your comments and add a dot `.` at the end so it looks like a proper sentence ? We all do that (except Christophe when he forgets) and it makes the comments look more professional. This obviously applies to all comments.

File: packages/plugin-link/src/Link.ts (hunk 9 above)
Comment (dmo-odoo) [reply packages/plugin-link/src/Link.ts:171]: This comment should not have been resolved as it is still relevant as of now. I'm unresolving it.

File: packages/plugin-char/src/Char.ts [hunk 15]
```diff
@@ -47,7 +47,12 @@ export class Char<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T>
         const range = params.context.range;
         const text = params.text;
         const inline = this.editor.plugins.get(Inline);
-        const modifiers = inline.getCurrentModifiers(range);
+        let modifiers = inline.getCurrentModifiers(range);
+
+        const previousSibling = range.start.previousSibling();
+        if (!previousSibling) {
```
Comment (dmo-odoo) [inline packages/plugin-char/src/Char.ts:53]: I was confused by this for a moment (even though I knew what the feature was about) so I suggest to add a comment on the line before `const previousSibling`:
```// Ony preserved modifiers are applied at the start of a container.```

File: packages/plugin-inline/src/Inline.ts [hunk 16]
```diff
@@ -147,7 +148,9 @@ export class Inline<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<
 
         let inlineToCopyModifiers: VNode;
         if (range.isCollapsed()) {
-            inlineToCopyModifiers = range.start.previousSibling() || range.start.nextSibling();
+            inlineToCopyModifiers =
+                range.start.previousSibling(node => !node.test(LineBreakNode)) ||
+                range.start.nextSibling();
```
Comment (dmo-odoo) [inline packages/plugin-inline/src/Inline.ts:153]: I'm not sure I understand this fix. What it does is that it takes the first previous sibling that is *not* a `LineBreakNode`, but why does it matter ? Besides, are there no tests for this fix or are they all put together in the "tests" commit ?

File: packages/plugin-layout/src/LayoutEngine.ts (hunk 12 above)
Comment (sebgeelen) [reply packages/plugin-layout/src/LayoutEngine.ts:11]: Nice to know, I will change it

File: packages/plugin-inline/src/Inline.ts (hunk 16 above)
Comment (sebgeelen) [reply packages/plugin-inline/src/Inline.ts:153]: This allow the `getCurrentModifiers()` method to return the same modifiers whether `this.cache.modifiers` is empty or not.
This fix a difference of behaviour between typing a character just after a `LineBreak` or after changing the cursor position the beginning of the line and typing a character.
In the latter, the modifiers was empty.

If this is not a issue, we can remove this commit.

File: packages/plugin-link/src/Link.ts (hunk 9 above)
Comment (sebgeelen) [reply packages/plugin-link/src/Link.ts:171]: The comment below is not clear enough ?

File: packages/plugin-link/src/Link.ts (hunk 14 above)
Comment (sebgeelen) [reply packages/plugin-link/src/Link.ts:172]: will do

File: packages/plugin-link/src/Link.ts (hunk 9 above)
Comment (dmo-odoo) [reply packages/plugin-link/src/Link.ts:171]: Only proper function doc written above the function in `/**` format is read by automated documentation systems like `tsdoc`. This is why we (try to) document all our functions in that way. You can open other files to see examples of it. Please note that we do not document the types of the parameters since this is redundant with the typing of Typescript which is correctly parsed by `tsdoc`.

File: packages/plugin-link/src/Link.ts (hunk 13 above)
Comment (Goaman) [reply packages/plugin-link/src/Link.ts:68]: Why would you want to use them as something else than props?
For me, props is good because it's the standard way of naming the parameters that a component receive.

Comment (dmo-odoo) [reply packages/plugin-link/src/Link.ts:68]: cc @sebgeelen for renaming the `params` parameter into `props` up along the entire function call chain.
